Mathlib/RingTheory/Coalgebra/TensorProduct.lean

```diff
@@ -1,90 +1,161 @@
-since it relies on the monoidal category structure on `ModuleCat R`, which is currently
-universe monomorphic. Any contribution that achieves universe polymorphism would be welcome. For
-instance, the tensor product of coalgebras in the
-[FLT repo](https://github.com/ImperialCollegeLondon/FLT/blob/eef74b4538c8852363936dfaad23e6ffba72eca5/FLT/mathlibExperiments/Coalgebra/TensorProduct.lean)
-is already universe polymorphic since it does not go via category theory.
+variable {R A B : Type*} [CommSemiring R] [AddCommMonoid B] [AddCommMonoid A]
+    [Module R A] [Module R B] [Coalgebra R A] [Coalgebra R B]
 
--/
+namespace TensorProduct
 
-universe v u
+open Coalgebra
 
-open CategoryTheory
-open scoped TensorProduct
+noncomputable
+instance instCoalgebraStruct : CoalgebraStruct R (A ⊗[R] B) where
+  comul :=
+    AlgebraTensorModule.tensorTensorTensorComm R R A A B B ∘ₗ
+      AlgebraTensorModule.map comul comul
+  counit := AlgebraTensorModule.rid R R R ∘ₗ AlgebraTensorModule.map counit counit
 
-section
+lemma comul_def :
+    Coalgebra.comul (R := R) (A := A ⊗[R] B) =
+      AlgebraTensorModule.tensorTensorTensorComm R R A A B B ∘ₗ
+        AlgebraTensorModule.map Coalgebra.comul Coalgebra.comul :=
+  rfl
 
-variable {R M N P Q : Type u} [CommRing R]
-  [AddCommGroup M] [AddCommGroup N] [Module R M] [Module R N] [Coalgebra R M] [Coalgebra R N]
-
-open MonoidalCategory in
-noncomputable instance TensorProduct.instCoalgebra : Coalgebra R (M ⊗[R] N) :=
-  let I := Monoidal.transport ((CoalgCat.comonEquivalence R).symm)
-  CoalgEquiv.toCoalgebra
-    (A := (CoalgCat.of R M ⊗ CoalgCat.of R N : CoalgCat R))
-    { LinearEquiv.refl R _ with
-      counit_comp := rfl
-      map_comp_comul := by
-        rw [CoalgCat.ofComonObjCoalgebraStruct_comul]
-        simp [-Mon_.monMonoidalStruct_tensorObj_X,
-          ModuleCat.MonoidalCategory.instMonoidalCategoryStruct_tensorHom,
-          ModuleCat.hom_comp, ModuleCat.of, ModuleCat.ofHom, comul_def,
-          ModuleCat.MonoidalCategory.tensorμ_eq_tensorTensorTensorComm] }
+@[deprecated (since := "2025-04-09")] alias instCoalgebraStruct_comul := comul_def
 
-end
+lemma counit_def :
+    Coalgebra.counit (R := R) (A := A ⊗[R] B) =
+      AlgebraTensorModule.rid R R R ∘ₗ AlgebraTensorModule.map counit counit :=
+  rfl
+
+@[deprecated (since := "2025-04-09")] alias instCoalgebraStruct_counit := counit_def
+
+@[simp]
+lemma comul_tmul (x : A) (y : B) :
+    comul (x ⊗ₜ y) =
+      AlgebraTensorModule.tensorTensorTensorComm R R A A B B (comul x ⊗ₜ comul y) := rfl
+
+@[simp]
+lemma counit_tmul (x : A) (y : B) :
+    counit (R := R) (x ⊗ₜ[R] y) = counit (R := R) y • counit (R := R) x := rfl
+
+/-- `expand_comul R x with x₁ x₂` attempts to replace `comul (R := R) x` by
+`x₁ ⊗ₜ x₂` via linearity. -/
+scoped macro "expand_comul" ring:term ", " var:term "with " var₁:ident var₂:ident : tactic =>
+  `(tactic|
+    (induction comul (R := $ring) $var with
```

Comment:
I guess this could be
```suggestion
scoped macro "tensor_induction " var:term "with " var₁:ident var₂:ident : tactic =>
  `(tactic|
    (induction $term with
```
and then `tensor_induction comul (R := R) var with x y` where you use it below. This is obviously a bit wordier, but nothing in the tactic is really about comul other than the major premise.

Reply:
Are you suggesting `hopf_tensor_induction comul (R := R) var with x y` or `hopf_tensor_induction R, var with x y`? And still scoped in `Coalgebra`?

Reply:
I am open to change to whatever syntax as long as it fits in one line. But I suspect we should first think about how to generalize the actual macro before generalizing the syntax (which I am happy to experiment on), but such a task is hopefully out of scope of this PR.